sources/csv/data.go

```diff
@@ -0,0 +1,284 @@
+				}
+				processDataRow(conv, table.Table_name, srcCols, values)
+			}
+		}
+	}
+	return nil
+}
+
+// processDataRow converts a row into go data types as per the client libs.
+func processDataRow(conv *internal.Conv, tableName string, srcCols []string, values []string) {
+	cvtVals, err := convertData(conv, tableName, srcCols, values)
+	if err != nil {
+		conv.Unexpected(fmt.Sprintf("Error while converting data: %s\n", err))
+		conv.StatsAddBadRow(tableName, conv.DataMode())
+		conv.CollectBadRow(tableName, srcCols, values)
+	} else {
+		conv.WriteRow(tableName, tableName, srcCols, cvtVals)
+	}
+}
+
+// convertData currently only supports scalar data types.
+func convertData(conv *internal.Conv, tableName string, srcCols []string, values []string) ([]interface{}, error) {
+	var v []interface{}
+	colDefs := conv.SpSchema[tableName].ColDefs
+	for i, val := range values {
+		colName := srcCols[i]
+		x, err := convScalar(colDefs[colName].T, val)
+		if err != nil {
+			return nil, err
+		}
+		v = append(v, x)
+	}
+	return v, nil
+}
+
+func convScalar(spannerType ddl.Type, val string) (interface{}, error) {
+	switch spannerType.Name {
+	case ddl.Bool:
+		return convBool(val)
+	case ddl.Bytes:
+		return convBytes(val)
+	case ddl.Date:
+		return convDate(val)
+	case ddl.Float64:
+		return convFloat64(val)
+	case ddl.Int64:
+		return convInt64(val)
+	case ddl.Numeric:
+		return convNumeric(val)
+	case ddl.String:
+		return val, nil
+	case ddl.Timestamp:
+		return convTimestamp(val)
+	case ddl.JSON:
+		return val, nil
+	default:
+		return val, fmt.Errorf("data conversion not implemented for type %v", spannerType)
+	}
+}
+
+func convBool(val string) (bool, error) {
+	b, err := strconv.ParseBool(val)
+	if err != nil {
+		return b, fmt.Errorf("can't convert to bool: %w", err)
+	}
+	return b, err
+}
+
+func convBytes(val string) ([]byte, error) {
+	// convert a string to a byte slice.
+	b := []byte(val)
+	return b, nil
+}
+
+func convDate(val string) (civil.Date, error) {
+	d, err := civil.ParseDate(val)
+	if err != nil {
+		return d, fmt.Errorf("can't convert to date: %w", err)
+	}
+	return d, err
+}
+
+func convFloat64(val string) (float64, error) {
+	f, err := strconv.ParseFloat(val, 64)
+	if err != nil {
+		return f, fmt.Errorf("can't convert to float64: %w", err)
+	}
+	return f, err
+}
+
+func convInt64(val string) (int64, error) {
+	i, err := strconv.ParseInt(val, 10, 64)
+	if err != nil {
+		return i, fmt.Errorf("can't convert to int64: %w", err)
+	}
+	return i, err
+}
+
+// convNumeric maps a source database string value (representing a numeric)
+// into a string representing a valid Spanner numeric.
+func convNumeric(val string) (interface{}, error) {
+	r := new(big.Rat)
+	if _, ok := r.SetString(val); !ok {
+		return "", fmt.Errorf("can't convert %q to big.Rat", val)
+	}
+	return r, nil
+}
+
+func convTimestamp(val string) (t time.Time, err error) {
+	t, err = time.Parse("2006-01-02 15:04:05", val)
```

Comment:
What is this timestamp ?

Reply:
This is the layout aka a template so the function knows the format. Thats how the function accepts it.